torch/_refs/__init__.py

```diff
@@ -3553,6 +3554,43 @@ def empty_strided(
     )
 
 
+@register_decomposition(torch.ops.aten.eye)
+@out_wrapper()
+def eye(
+    n: int,
+    m: Optional[int] = None,
+    *,
+    dtype: Optional[torch.dtype] = None,
+    layout: torch.layout = torch.strided,
+    device: Optional[torch.device] = None,
+    pin_memory: bool = False,
+    requires_grad: bool = False,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.eye
+    """
+    # TODO: no support for layout, pin_memory
+    assert layout == torch.strided
+    assert pin_memory is False
+
+    if m is None:
+        m = n
+
+    check(n >= 0, lambda: f"n must be greater or equal to 0, got {n}")
+    check(m >= 0, lambda: f"m must be greater or equal to 0, got {m}")
+
+    range_n = torch.arange(n, dtype=torch.int64, device=device, requires_grad=False)
+    range_m = torch.arange(m, dtype=torch.int64, device=device, requires_grad=False)
+
+    cond = range_n.unsqueeze(-1) == range_m
+    # TODO: pin_memory=pin_memory, layout=layout
+    one = torch.ones(1, dtype=dtype, device=device, requires_grad=requires_grad)
+    zero = torch.zeros(1, dtype=dtype, device=device, requires_grad=requires_grad)
```

Comment:
fwiw, this isn't going to differentiate correctly (this doesn't matter atm because we don't differentiate through refs); eye is supposed to produce a leaf node.

Reply:
@ezyang, @nkaretnikov  would it work if `requires_grad` is set after the tensor is created?

```
    one = torch.ones(1, dtype=dtype, device=device, requires_grad=False)
    zero = torch.zeros(1, dtype=dtype, device=device, requires_grad=False)
    result = torch.where(cond, one, zero)
    result.requires_grad_(requires_grad)
```

Reply:
re `requires_grad_`: isn't this an anti-primtorch thing to do as we're trying to avoid inplace ops?

Reply:
I believe that `requires_grad_` might be an exception, because it does not modify the contents of a tensor. Though, I am not 100% sure if it is safe to use it in primtorch.

Reply:
What about 
```
dtype = dtype or torch.get_default_dtype()
result = cond.to(dtype) #or convert_data_type, whatever primtorch has for dtype conversion
```
And yeah factory functions (like full_like, zeros_like etc) are doing it wrong now, and aren't returning leaf node like they should, calling `requires_grad_` on result seems like a better solution.